File: Graph Algorithms/Building Teams/building_teams_keshav.cpp
#include<bits/stdc++.h>
#include <sys/resource.h>
#define INF 1000000000
#define endl '\n'
const double PI = 3.141592653589793238460;
const int mxN = 2e5;
using namespace std;

class Vertex{
    public:
    int data;
    int visited;
    int group;
    vector<int> connections;
};

void bfs(unordered_map<int, Vertex>& graph, int i, int color){
    vector<int> q;
    q.push_back(i);
    while(q.size() != 0){
        vector<int> new_q;
        for(int j : q){
            if(graph[j].visited) continue;
            graph[j].visited = 1;
            for(int k: graph[j].connections){
                new_q.push_back(k);
            }
            graph[j].group = color;
        }
        color ^= 1;
        q = new_q;
    }
}
int main(){
    rlimit cpu_time{.rlim_cur = 1, .rlim_max = 5}; 
    setrlimit(RLIMIT_CPU, &cpu_time);
    ios_base::sync_with_stdio(0);
    int n, m,a,b;
    cin >> n >> m;
    unordered_map<int , Vertex> graph;
    for(int i = 0; i < n; i++){
        graph[i+1] = Vertex();
        graph[i+1].data = i+1;
        graph[i+1].visited = 0;
        graph[i+1].group = 0;
    }
    for(int i = 0 ; i < m; i++){
        cin >> a >> b;
        graph[a].connections.push_back(b);
        graph[b].connections.push_back(a);
    }
    for(int i = 1; i <=n ; i++){
        if(graph[i].visited) continue;
        bfs(graph,i,0);
    }
    for(int i = 1; i <= n; i++){
        for(int j:graph[i].connections){
            if(graph[i].group == graph[j].group){
                cout << "IMPOSSIBLE" << endl;
                return 0;
            }
        }
    }
    for(int i = 1; i <=n; i++){
        cout << graph[i].group + 1 << " ";
    }
    cout << endl;
}

File: Graph Algorithms/Building Teams/building_teams_keshav.py
from collections import deque
class Vertex:
    def __init__(self,data):
        self.data = data 
        self.visited = False
        self.group = 1 
        self.connections = []

def bfs(graph,i,color):
    q = deque()
    q.append(i)
    while q:
        new = deque()
        for j in q:
            vertex = graph[j]
            if vertex.visited: continue
            vertex.visited = True 
            vertex.group = color
            for k in vertex.connections:
                new.appendleft(k)
        color ^=1 
        q = new

def solve():
    n, m = map(int,input().split())
    graph = dict()
    for i in range(1,n+1):
        graph[i] = Vertex(i)
    for _ in range(m):
        a,b = map(int,input().split())
        graph[a].connections.append(b)
        graph[b].connections.append(a)
    # color all the vertices levelwise using bfs
    for i in range(1,n+1):
        if graph[i].visited: continue
        bfs(graph,i,0)
    
    # check if there are adjacent edges with the same color for 
    # impossible condition 
    for i in range(1,n+1):
        for j in graph[i].connections:
            if graph[i].group == graph[j].group:
                print('IMPOSSIBLE') 
                return 
    
    for i in range(1,n+1):
        print(graph[i].group + 1, end = " ")
    print()
        
if __name__ == '__main__':
    solve() 
